src/bin/enclave-runc-wrapper.rs

```diff
@@ -0,0 +1,241 @@
+use k8s_openapi::api::core::v1;
+use std::convert::TryFrom;
+
+// TODO: To be used for cri-o.
+// static ANNOTATION_K8S_LABELS: &str = "io.kubernetes.cri-o.Labels";
+
+// static LABEL_NAMESPACE: &str = "io.kubernetes.pod.namespace";
+static LABEL_POLICY_ENFORCE: &str = "pod-security.kubernetes.io/enforce";
+// static LABEL_POLICY_AUDIT: &str = "pod-security.kubernetes.io/audit";
+// static LABEL_POLICY_WARN: &str = "pod-security.kubernetes.io/warn";
+
+static ANNOTATION_CONTAINERD_LOG_DIRECTORY: &str = "io.kubernetes.cri.sandbox-log-directory";
+static ANNOTATION_CONTAINERD_SANDBOX_ID: &str = "io.kubernetes.cri.sandbox-id";
+
+#[derive(thiserror::Error, Debug)]
+enum NewContainerError {
+    #[error("could not retrieve the runc status")]
+    Status(#[from] std::io::Error),
+
+    #[error("could not format")]
+    Format(#[from] std::fmt::Error),
+
+    #[error("could not convert bytes to utf-8 string")]
+    Utf8(#[from] std::string::FromUtf8Error),
+
+    #[error("could not parse JSON")]
+    Json(#[from] serde_json::Error),
+}
+
+fn container_namespace(
+    container_id: &str,
+    container_root: Option<std::string::String>,
+) -> Result<Option<std::string::String>, NewContainerError> {
+    let mut state_cmd = std::process::Command::new("runc");
+    if let Some(ref v) = container_root {
+        state_cmd.arg("--root").arg(v);
+    }
+
+    state_cmd.arg("state").arg(container_id);
+    let state_raw = String::from_utf8(state_cmd.output()?.stdout)?;
+
+    let state: serde_json::Value = serde_json::from_str(&state_raw)?;
+    let annotations_o = state["annotations"].as_object();
+
+    match annotations_o {
+        Some(annotations) => {
+            // containerd
+            if annotations.contains_key(ANNOTATION_CONTAINERD_LOG_DIRECTORY) {
+                // containerd doesn't expose k8s namespaces directly. They have
+                // to be parsed from the log directory path, where the first
+                // part of the filename is the namespace.
+                let log_directory = annotations[ANNOTATION_CONTAINERD_LOG_DIRECTORY]
+                    .as_str()
+                    .unwrap();
+                let p = std::path::PathBuf::from(log_directory);
+                let file_name = p.file_name().unwrap().to_str().unwrap();
+                let mut splitter = file_name.split('_');
+                let namespace = splitter.next().unwrap().to_string();
+
+                return Ok(Some(namespace));
+            } else if annotations.contains_key(ANNOTATION_CONTAINERD_SANDBOX_ID) {
+                // When a container is running as a part of a previously created
+                // pod, the log directory path has to be retrieved from the
+                // sandbox container.
+                let sandbox_id = annotations[ANNOTATION_CONTAINERD_SANDBOX_ID]
+                    .as_str()
+                    .unwrap();
+                return container_namespace(sandbox_id, container_root);
```

Comment:
Are we concerned about unwrap panics?

Reply:
I was thinking if I should use `ok_or_error` instead of `unwrap`, but on the other hand, I'm not sure if it's worth to write all those new errors manually.

After all, the panic here is very unlikely to happen - I'm unwrapping only when the key exists.

Reply:
I agree that it's unlikely. We should be good.